### PR TITLE
new-app should chase a defaulted "latest" tag to the stable ref

### PR DESCRIPTION
**Pull request by smarterclayton**
If an image stream has "latest" with an ImageStreamTag pointing to
another tag in the stream, and a user requests that stream via oc
new-app <stream>, new-app should create the app pointing to the
destination tag instead of "latest".  I.e.:

    # is mysql has three tags, latest->5.6, 5.6, and 5.6
    oc new-app mysql
    # create app pointing to mysql:5.6

Guarantees the admin can provide stable tags and the user ends up using a stable image stream.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@danmcp the last thing we needed to add from our original "openshift next" design.

**Comment by smarterclayton** [comment]
@bparees fixed an issue in template search, a suggested name bug (for where we were suggesting the wrong app name on an existing image stream), and implemented the ref chase.

**Comment by smarterclayton** [comment]
[test]

File: pkg/generate/app/componentresolvers.go [hunk 1]
```diff
@@ -39,41 +41,36 @@ type PerfectMatchWeightedResolver []WeightedResolver
...
-			sort.Sort(ScoredComponentMatches(inexact))
-			if inexact[0].Score == 0.0 && (len(inexact) == 1 || inexact[1].Score != 0.0) {
-				return inexact[0], nil
-			}
-			for _, m := range inexact {
+			for _, m := range matches {
 				if resolver.Weight != 0.0 {
 					m.Score = resolver.Weight * m.Score
 				}
 				imperfect = append(imperfect, m)
```
Comment (bparees) [inline pkg/generate/app/componentresolvers.go]: this name is confusing.  they are not necessarily imperfect matches, you could have a bunch of exact(perfect) matches in this array.

Comment (smarterclayton) [inline pkg/generate/app/componentresolvers.go]: I may rename to "candidates"

Comment (bparees) [inline pkg/generate/app/componentresolvers.go]: +1

File: pkg/generate/app/imagestreamlookup.go [hunk 2]
```diff
@@ -33,9 +33,11 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
 		if len(ref.Namespace) != 0 {
 			namespaces = []string{ref.Namespace}
 		}
+		followTag := false
```
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: comment this. i have no idea what a "followTag" is.

File: pkg/generate/app/imagestreamlookup.go [hunk 3]
```diff
@@ -33,9 +33,11 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
...
 		searchTag := ref.Tag
 		if len(searchTag) == 0 {
 			searchTag = imageapi.DefaultImageTag
+			followTag = true
```
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: comment this.  on what basis did we decide to set followTag to true?  what user intent are we trying to execute on here?  ("user didn't specify a specific tag, therefore we are going to try to match to....")

File: pkg/generate/app/imagestreamlookup.go [hunk 4]
```diff
@@ -56,10 +58,20 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
 				} else {
 					imageref, _ := imageapi.ParseDockerImageReference(term)
 					imageref.Name = stream.Name
+
+					localSearchTag := searchTag
```
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: why is this variable needed?  the original searchTag value is never referenced again after you assign localSearchTag below, so you could have just continued using searchTag.

File: pkg/generate/app/imagestreamlookup.go [hunk 5]
```diff
@@ -56,10 +58,20 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
...
+					if specTag, ok := stream.Spec.Tags[searchTag]; ok && followTag {
```
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: // "if the user didn't specify a specific tag (so we're using latest), and the "latest" imagestreamtag is actually just 
// following another imagestreamtag, then use the actual imagestreamtag that "latest" is following as the match 
// result instead of the "latest"imagestream tag".  Also only do this if the name IST that "latest" is following does 
// not contain a ":" because....???

at least I think that's what this is doing, i'd like to save the next person who hits this code the trouble of trying to figure it out.

File: pkg/generate/app/templatelookup.go [hunk 6]
```diff
@@ -57,9 +61,13 @@ func (r TemplateSearcher) Search(terms ...string) (ComponentMatches, error) {
 						Score:       score,
 						Template:    template,
 					})
+					break
 				}
 			}
 		}
+		if exact {
+			break
```
Comment (bparees) [inline pkg/generate/app/templatelookup.go]: shouldn't this be checking StopOnMatch?

File: pkg/generate/app/templatelookup.go [hunk 7]
```diff
@@ -21,34 +21,38 @@ type TemplateSearcher struct {
 	Client                    client.TemplatesNamespacer
 	TemplateConfigsNamespacer client.TemplateConfigsNamespacer
 	Namespaces                []string
+	StopOnMatch               bool
```
Comment (bparees) [inline pkg/generate/app/templatelookup.go]: StopOnMatch or StopOnExactMatch?

if it's the latter, rename it or comment it.

File: pkg/generate/app/templatelookup.go [hunk 8]
```diff
@@ -57,9 +61,13 @@ func (r TemplateSearcher) Search(terms ...string) (ComponentMatches, error) {
 						Score:       score,
 						Template:    template,
 					})
+					break
```
Comment (bparees) [inline pkg/generate/app/templatelookup.go]: i'm confused by this break.  what if the term you "matched" an item to was not an exact match?  don't you want to search on the other terms to see if there is a better match?  (i'm not clear under what conditions we'd actually have multiple terms either, but...)

File: pkg/generate/app/templatelookup.go [hunk 9]
```diff
@@ -92,7 +100,8 @@ func (r *TemplateFileSearcher) Search(terms ...string) (ComponentMatches, error)
 			Object()
 
 		if err != nil {
-			return nil, err
+			glog.V(5).Infof("Error from template search: %v", err)
+			continue
```
Comment (bparees) [inline pkg/generate/app/templatelookup.go]: not aborting on an error seems like a mistake. or at least printing it as a warning.  Doesn't this mean if the template file the user intended to use has a typo, you're going to silently skip over that file?

**Comment by bparees** [comment]
@smarterclayton review complete.

File: pkg/generate/app/imagestreamlookup.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/generate/app/imagestreamlookup.go]: Clarity of intent - localSearchTag is not the same thing as a searchTag,
can be a better name though.

File: pkg/generate/app/templatelookup.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/generate/app/templatelookup.go]: exact is never set unless StopOnMatch is set.

File: pkg/generate/app/templatelookup.go (hunk 8 above)
Comment (smarterclayton) [inline pkg/generate/app/templatelookup.go]: If we had multiple terms we'd stop at the first level that had matches.
Not on the first match.

File: pkg/generate/app/templatelookup.go (hunk 9 above)
Comment (smarterclayton) [inline pkg/generate/app/templatelookup.go]: Hrm, maybe.  I'm not happy with the error handling here.

File: pkg/generate/app/templatelookup.go (hunk 6 above)
Comment (bparees) [inline pkg/generate/app/templatelookup.go]: I don't see any logic that's even using StopOnMatch in this PR.  what am i missing?

File: pkg/generate/app/imagestreamlookup.go (hunk 4 above)
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: near as i can tell the intent is "this is either the original searchTag provided by the user, or it is the Tag which is followed by the :latest tag", which seems hard to convey in a name, so maybe a comment would be better.

File: pkg/generate/app/templatelookup.go (hunk 8 above)
Comment (bparees) [inline pkg/generate/app/templatelookup.go]: is that what you really want?  I guess i don't understand what the "levels" (multiple terms) are here...  but i would think if term1 had a lousy match, you'd still want to check to see if term2 had a good/better match.

File: pkg/generate/app/templatelookup.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/generate/app/templatelookup.go]: Nothing, it was a bug

**Comment by smarterclayton** [comment]
The last commit is a doozy but does greatly simplify the handling of errors.  I'm not convinced our error handling has enough tests, so I need to go through and add more multiple failure conditions and validate them.

**Comment by smarterclayton** [comment]
Needs two fixes:

* Changes to returning errors means that DockerHub v2 fails out on an image that doesn't exist (because docker hub doesn't distinguish between this image exists and you don't have auth), so we'll need something new there
* Multiple error output needs to be thought through more - we probably want to coalesce errors of types into chunks.

File: pkg/generate/app/componentresolvers.go [hunk 10]
```diff
@@ -38,51 +41,47 @@ type PerfectMatchWeightedResolver []WeightedResolver
...
-			sort.Sort(ScoredComponentMatches(inexact))
-			if inexact[0].Score == 0.0 && (len(inexact) == 1 || inexact[1].Score != 0.0) {
-				return inexact[0], nil
-			}
-			for _, m := range inexact {
-				if resolver.Weight != 0.0 {
-					m.Score = resolver.Weight * m.Score
-				}
-				imperfect = append(imperfect, m)
+				m.Score = resolver.Weight * m.Score
```
Comment (bparees) [inline pkg/generate/app/componentresolvers.go]: though it doesn't really matter, this would be more readable if you used groupWeight instead of resolver.Weight here and above.

File: pkg/generate/app/componentresolvers.go [hunk 11]
```diff
@@ -90,58 +89,22 @@ func (r PerfectMatchWeightedResolver) Resolve(value string) (*ComponentMatch, er
 					}
 				}
 			}
-		}
-		return nil, ErrNoMatch{value: value}
+		}*/
```
Comment (bparees) [inline pkg/generate/app/componentresolvers.go]: is this getting uncommented prior to merge?  or hoping for @rhcarvalho's changes?

File: pkg/generate/app/componentresolvers.go [hunk 12]
```diff
@@ -197,21 +154,18 @@ type HighestUniqueScoreResolver struct {
 // Resolve resolves as the highest scored match returned by the Searcher, and
 // guarantees the match is unique (the only match with that given score)
 func (r HighestUniqueScoreResolver) Resolve(value string) (*ComponentMatch, error) {
```
Comment (bparees) [inline pkg/generate/app/componentresolvers.go]: do we really have use cases for all these resolvers?  (first, highest, highest unique)
seems like "highest score" by itself ought to be sufficient.

File: pkg/generate/app/dockerimagelookup.go [hunk 13]
```diff
@@ -41,13 +40,18 @@ type DockerClientSearcher struct {
 }
 
 // Search searches all images in local docker server for images that match terms
-func (r DockerClientSearcher) Search(terms ...string) (ComponentMatches, error) {
+func (r DockerClientSearcher) Search(precise bool, terms ...string) (ComponentMatches, []error) {
 	componentMatches := ComponentMatches{}
 	errs := []error{}
 	for _, term := range terms {
+		if term == "__dockerimage_fail" {
```
Comment (bparees) [inline pkg/generate/app/dockerimagelookup.go]: debugging?

File: pkg/generate/app/imagestreamlookup.go [hunk 14]
```diff
@@ -18,35 +18,45 @@ type ImageStreamSearcher struct {
...
 	componentMatches := ComponentMatches{}
+	var errs []error
 	for _, term := range terms {
 		ref, err := imageapi.ParseDockerImageReference(term)
 		if err != nil || len(ref.Registry) != 0 {
-			return nil, fmt.Errorf("image streams must be of the form [<namespace>/]<name>[:<tag>|@<digest>]")
+			glog.V(2).Infof("image streams must be of the form [<namespace>/]<name>[:<tag>|@<digest>], term %q did not qualify", term)
+			continue
+		}
+		if term == "__imagestream_fail" {
```
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: debugging?

File: pkg/generate/app/imagestreamlookup.go [hunk 15]
```diff
@@ -56,28 +66,41 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
...
+				imageref := original
 				imageref.Name = stream.Name
 				imageref.Registry = ""
 				matchName := fmt.Sprintf("%s/%s", stream.Namespace, stream.Name)
 
-				latest := imageapi.LatestTaggedImage(stream, searchTag)
+				// When an image stream contains a tag that references another local tag, and the user has not
+				// provided a tag themselves (i.e. they asked for mysql and we defaulted to mysql:latest), walk
+				// the chain of references to the end. This ensures that applications can default to using a "stable"
+				// branch by giving the control over version to the image stream author.
```
Comment (bparees) [inline pkg/generate/app/imagestreamlookup.go]: this comment brings me great joy.

**Comment by rhcarvalho** [comment]
Maybe [new-app should chase a defaulted "latest" tag to the stable ref](https://github.com/smarterclayton/origin/commit/01432fe4edce8619c14454960a64222ccb207a39) and [Add a test for chasing references in new-app](https://github.com/smarterclayton/origin/commit/a10996f2bedb147174bf3d4b4fcf900fa4dcc5d2) should be squashed together?

File: pkg/generate/app/imagestreamlookup.go [hunk 16]
```diff
@@ -278,19 +301,32 @@ func (r *ImageStreamByAnnotationSearcher) annotationMatches(stream *imageapi.Ima
...
 	for _, namespace := range r.Namespaces {
 		streams, err := r.getImageStreams(namespace)
 		if err != nil {
-			return nil, err
+			errs = append(errs, err)
+			continue
 		}
 		for i := range streams {
 			for _, term := range terms {
+				if term == "__imagestreamannotation_fail" {
```
Comment (rhcarvalho) [inline pkg/generate/app/imagestreamlookup.go]: debugging?

**Comment by smarterclayton** [comment]
Ok, this is updated with logic to unify the errors.  I went through and tried to whitelist certain error conditions as "known not found".  The docker hub one is the hardest, because DockerHub gives us Unauthorized for "NotFound' in some cases, and that's not a great experience.

File: pkg/generate/app/componentresolvers.go (hunk 11 above)
Comment (smarterclayton) [inline pkg/generate/app/componentresolvers.go]: It's not needed any more - you see the error along with no match

File: pkg/generate/app/componentresolvers.go (hunk 12 above)
Comment (smarterclayton) [inline pkg/generate/app/componentresolvers.go]: Maybe not, but I don't want to introduce even more churn in this PR.  They're in the resolver chain today for the other types

File: pkg/generate/app/dockerimagelookup.go (hunk 13 above)
Comment (smarterclayton) [inline pkg/generate/app/dockerimagelookup.go]: More like something concrete we can use in test cases.  I occasionally like being able to make something fail concretely.

File: pkg/generate/app/componentresolvers.go (hunk 11 above)
Comment (smarterclayton) [inline pkg/generate/app/componentresolvers.go]:
```
$ oc new-app foo.json
error: unable to load template file "foo.json": invalid character '}' after object key
error: no match for "foo.json"

The 'oc new-app' command will match arguments to the following types:

  1. Images tagged into image streams in the current project or the 'openshift' project
     - if you don't specify a tag, we'll add ':latest'
  2. Images in the Docker Hub, on remote registries, or on the local Docker engine
  3. Templates in the current project or the 'openshift' project
  4. Git repository URLs or local paths that point to Git repositories

--allow-missing-images can be used to point to an image that does not exist yet.

See 'oc new-app -h' for examples.
```

Will add a test explicitly for this.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Stabs bash repeatedly in the neck and then goes to sleep.

**Comment by smarterclayton** [comment]
Zomg

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
I don't believe this fixed the import flakes

[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
ZOMG

**Comment by smarterclayton** [comment]
Four non flaky runs in a row - let's try to close this review out tomorrow and maybe we'll be better off

**Comment by smarterclayton** [comment]
[test]

File: test/cmd/newapp.sh [hunk 17]
```diff
@@ -9,10 +9,18 @@ source "${OS_ROOT}/hack/util.sh"
 source "${OS_ROOT}/hack/cmd_util.sh"
 os::log::install_errexit
 
+# Cleanup cluster resources created by this test
+(
+  set +e
+  oc delete all,templates --all
+  exit 0
+) &>/dev/null
```
Comment (rhcarvalho) [inline test/cmd/newapp.sh]: Is this supposed to be in the final patch?
Why do we delete resources prior to running the tests?

File: test/cmd/newapp.sh [hunk 18]
```diff
@@ -126,6 +158,10 @@ os::cmd::expect_success 'oc delete template ruby-helloworld-sample'
 os::cmd::expect_success_and_text 'oc new-app mysql --name=db' 'db'
 os::cmd::expect_success 'oc new-app https://github.com/openshift/ruby-hello-world -l app=ruby'
 os::cmd::expect_success 'oc delete all -l app=ruby'
+# check for error when template JSON file has errors
+jsonfile="${TMPDIR:-/tmp}/foo.json"
+echo '{"test"}' > "${jsonfile}"
```
Comment (rhcarvalho) [inline test/cmd/newapp.sh]: I searched for `TMPDIR` in the project, but didn't find any place that would set it to anything other than `/tmp`.
I'm assuming this file will not be deleted by the test framework after run, is it to be considered a problem?

File: test/cmd/newapp.sh [hunk 19]
```diff
@@ -126,6 +158,10 @@ os::cmd::expect_success 'oc delete template ruby-helloworld-sample'
...
+os::cmd::expect_failure_and_text "oc new-app '${jsonfile}'" "error: unable to load template file \"${jsonfile}\": invalid character '}' after object key"
```
Comment (rhcarvalho) [inline test/cmd/newapp.sh]: In practice, don't you think users will miss this error message when it is followed by a block of 10+ lines telling how new-app works?
I believe people may read the new-app instructions once, and in all subsequent runs our brains just pattern-matches the block of text and ignore it.

File: test/cmd/newapp.sh (hunk 17 above)
Comment (smarterclayton) [inline test/cmd/newapp.sh]: So we can rerun them as an end user - the original point of all the
individual scripts is that you can re run them over and over.

File: test/cmd/newapp.sh (hunk 18 above)
Comment (smarterclayton) [inline test/cmd/newapp.sh]: I don't expect it is - we simply have to have a file on disk somewhere.  I
may just move this to a fixture.

File: test/cmd/newapp.sh (hunk 19 above)
Comment (smarterclayton) [inline test/cmd/newapp.sh]: In practice we print:

$ commandline
<chunk of errors>
<chunk of text>

My concern about

$ commandline
<title of one error>
<chunk of text>
<chunk of errors>

is that the latter is also hard to parse.

File: test/cmd/newapp.sh (hunk 17 above)
Comment (rhcarvalho) [inline test/cmd/newapp.sh]: The pattern I've seen before in these tests was to do clean ups only when necessary between test cases. The "test framework" automatically creates a project for each `test/cmd/*.sh` file and after the tests run it will remove the whole project.

So the way I've been running these tests locally during development is:

```console
$ hack/test-cmd.sh test/cmd/newapp.sh
```

Perhaps you're running them in a smarter way?

What I don't like in the way I run it is that it takes a way too long time to start a server in every test run, making developing those tests quite painful. I often comment out all other tests in a file.

@smarterclayton could you describe how you develop and run test/cmd tests locally? I'd happily add the knowledge to [HACKING.md](https://github.com/openshift/origin/blob/master/HACKING.md#integration-tests).

File: test/cmd/newapp.sh (hunk 19 above)
Comment (rhcarvalho) [inline test/cmd/newapp.sh]: If you have a file in the current directory named `template.json` and run `oc new-app template.json`, what's the likelihood you meant a Docker image called "template.json"?

I think in that case new-app should just say the file has invalid JSON (+ details from the parser), and print nothing else.

File: test/cmd/newapp.sh (hunk 17 above)
Comment (smarterclayton) [inline test/cmd/newapp.sh]: What I'm doing is starting up the master in another cmd shell (openshift
start master --loglevel=5) and then just running test/cmd/newapp.sh.  A few
tests need preconditions, but any test that needs preconditions should set
them up anyway (for test-cmd).

File: test/cmd/newapp.sh (hunk 19 above)
Comment (smarterclayton) [inline test/cmd/newapp.sh]: Hrm - that would mean the template file resolver has to have top weight,
which means a local file will *always* take precedence over any image from
anywhere.  So having a file called php in the local directory would prevent
you from being able to create php things.  I agree with the statement, but
I think we'd have to only look at things with .yaml and .json extensions to
give them that weight (php.json I can buy being weighted above everything
else, php i can't).

Comment (rhcarvalho) [inline test/cmd/newapp.sh]: I agree with you @smarterclayton :+1:

File: test/cmd/newapp.sh (hunk 17 above)
Comment (rhcarvalho) [inline test/cmd/newapp.sh]: I could document that during development one could run `test/cmd/*.sh` files to test against a running server.

However, this also means that running a test might accidentally delete resources you had in the server.

Could we protect this clean up somehow? Say, only delete resources if an environment variable `OPENSHIFT_TEST_CMD_FORCE_CLEANUP` is set.
Then, by default, running a test would not delete resources in whatever namespace/project you happen to be.

Comment (smarterclayton) [inline test/cmd/newapp.sh]: That's possible.  But you could also accidentally delete things by running
oc delete all --all in normal use, or run rm -rf *.  At a certain level,
running tests accepts a risk.

Comment (smarterclayton) [inline test/cmd/newapp.sh]: Pushed an update to move the test json into fixtures.

Comment (rhcarvalho) [inline test/cmd/newapp.sh]: > But you could also accidentally delete things by running
> oc delete all --all in normal use,

I've hit this while working on the project :smile: . The catch here is that whenever the development server is restarted, it sends you back to the `default` namespace. https://github.com/openshift/origin/issues/2516

I "learned" to always include a "-n" argument in my `oc delete` calls.


> or run rm -rf *

I wouldn't like a project that deletes my files when I run the test suite. This reminds of the epic [bumblebee](https://github.com/MrMEEE/bumblebee-Old-and-abbandoned/commit/a047be85247755cdbe0acce6f1dafc8beb84f2ac#diff-3fbb47e318cd8802bd325e7da9aaabe8L351).


I wish there were no "risks" involved when running tests.
IMHO, it's easier to tolerate test suites that leaves behind temporary files (as long as it doesn't make your disk bleed) than having files changed or removed.

Comment (smarterclayton) [inline test/cmd/newapp.sh]: If we didn't put the registry or kubernetes service into the default
namespace, it would be safer.  That's been under discussion for awhile.

**Comment by smarterclayton** [comment]
Crap, new flake:

```

--- FAIL: TestImageStreamImportDockerHub-2 (15.00s)
	imageimporter_test.go:507: unexpected repository:
&api.RepositoryImportStatus{Status:unversioned.Status{TypeMeta:unversioned.TypeMeta{Kind:"",
APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"",
ResourceVersion:""}, Status:"Failure", Message:"Internal error
occurred: Get https://registry-1.docker.io/v2/: read tcp
52.72.23.195:443: use of closed network connection",
Reason:"InternalError",
Details:(*unversioned.StatusDetails)(0xc208609c00), Code:500},
Images:[]api.ImageImportStatus(nil), AdditionalTags:[]string(nil)}
	imageimporter_test.go:514: unexpected object: (*api.Image)(nil)
	imageimporter_test.go:518: unexpected object: (*api.Image)(nil)
	imageimporter_test.go:522: unexpected object: (*api.Image)(nil)
	imageimporter_test.go:526: unexpected object:
api.ImageImportStatus{Tag:"",
Status:unversioned.Status{TypeMeta:unversioned.TypeMeta{Kind:"",
APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"",
ResourceVersion:""}, Status:"Failure", Message:"Internal error
occurred: Get https://registry-1.docker.io/v2/: read tcp
52.72.23.195:443: use of closed network connection",
Reason:"InternalError",
Details:(*unversioned.StatusDetails)(0xc208609a80), Code:500},
Image:(*api.Image)(nil)}
FAIL

```

​
[test]

File: pkg/cmd/cli/cmd/newbuild.go [hunk 20]
```diff
@@ -204,33 +206,46 @@ func handleBuildError(c *cobra.Command, err error, fullName string) error {
...
-	}
-	switch err {
-	case newcmd.ErrNoInputs:
-		// TODO: suggest things to the user
-		return cmdutil.UsageError(c, newBuildNoInput, fullName)
-	default:
-		return err
-	}
+		groups.Add(
+			"no-matches",
```
Comment (bparees) [inline pkg/cmd/cli/cmd/newbuild.go]: i'm not seeing how this handling is different from the newapp.transformError() implementation?

File: pkg/generate/app/componentresolvers.go [hunk 21]
```diff
@@ -197,21 +139,18 @@ type HighestUniqueScoreResolver struct {
...
 	case 0:
-		return nil, ErrNoMatch{value: value}
+		return nil, ErrNoMatch{Value: value, Errs: err}
 	case 1:
-		return matches[0], nil
+		return matches[0], errors.NewAggregate(err)
 	default:
 		if matches[0].Score == matches[1].Score {
-			return nil, ErrMultipleMatches{value, matches}
+			return nil, ErrMultipleMatches{Value: value, Matches: matches, Errs: err}
```
Comment (bparees) [inline pkg/generate/app/componentresolvers.go]: should this only return(in the error object) the matches that have the same score, rather than all the matches?

File: pkg/generate/app/dockerimagelookup.go [hunk 22]
```diff
@@ -41,13 +40,18 @@ type DockerClientSearcher struct {
...
+			errs = append(errs, fmt.Errorf("unable to find the specified docker image: %s", term))
+			continue
+		}
+
 		ref, err := imageapi.ParseDockerImageReference(term)
 		if err != nil {
-			return nil, err
+			continue
```
Comment (bparees) [inline pkg/generate/app/dockerimagelookup.go]: continue?  shouldn't this at least aggregate the error?

File: pkg/generate/app/dockerimagelookup.go [hunk 23]
```diff
@@ -94,6 +95,10 @@ func (r DockerClientSearcher) Search(terms ...string) (ComponentMatches, error)
 		}
 		sort.Sort(termMatches)
 
+		if r.Client == nil {
+			continue
```
Comment (bparees) [inline pkg/generate/app/dockerimagelookup.go]: this doesn't look right...  r.Client was already used on line 76 (assuming no exact matches in the registry)

File: pkg/generate/app/dockerimagelookup.go [hunk 24]
```diff
@@ -161,9 +164,14 @@ type ImageImportSearcher struct {
 	Fallback      Searcher
 }
 
-func (s ImageImportSearcher) Search(terms ...string) (ComponentMatches, error) {
+func (s ImageImportSearcher) Search(precise bool, terms ...string) (ComponentMatches, []error) {
```
Comment (bparees) [inline pkg/generate/app/dockerimagelookup.go]: Godoc

File: pkg/generate/app/errors.go [hunk 25]
```diff
@@ -4,77 +4,67 @@ import (
...
 	}
-	return fmt.Sprintf("no match for %q, specify --allow-missing-images to use this image name.", e.value)
+	return fmt.Sprintf("no match for %q", e.Value)
 }
 
 // UsageError is the usage error message returned when no match is found.
-func (e ErrNoMatch) UsageError(commandName string) string {
-	return fmt.Sprintf("%[3]s - does a Docker image with that name exist?", e.value, commandName, e.Error())
+func (e ErrNoMatch) Suggestion(commandName string) string {
+	return fmt.Sprintf("%[3]s - does a Docker image with that name exist?", e.Value, commandName, e.Error())
```
Comment (bparees) [inline pkg/generate/app/errors.go]: shouldn't this be suggesting --allow-missing-images?

File: pkg/cmd/cli/cmd/newbuild.go (hunk 20 above)
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newbuild.go]: 3 suggestions vs 4

File: pkg/generate/app/componentresolvers.go (hunk 21 above)
Comment (smarterclayton) [inline pkg/generate/app/componentresolvers.go]: yes

File: pkg/generate/app/dockerimagelookup.go (hunk 22 above)
Comment (smarterclayton) [inline pkg/generate/app/dockerimagelookup.go]: passing a term that is not a valid docker image reference is not an error, it is simply "there is nothing that could possibly match this".  If we didn't parse the reference, or had partial matching, we could.

File: pkg/generate/app/dockerimagelookup.go (hunk 23 above)
Comment (smarterclayton) [inline pkg/generate/app/dockerimagelookup.go]: Will move it up above 70

File: pkg/generate/app/errors.go (hunk 25 above)
Comment (smarterclayton) [inline pkg/generate/app/errors.go]: newapp and newbuild are now responsible for that

File: pkg/cmd/cli/cmd/newbuild.go (hunk 20 above)
Comment (bparees) [inline pkg/cmd/cli/cmd/newbuild.go]: ah right.

File: pkg/generate/app/dockerimagelookup.go (hunk 22 above)
Comment (bparees) [inline pkg/generate/app/dockerimagelookup.go]: as a user i feel like i'd want to be told that, rather than repeatedly being told "nothing matched your stupid typo"

Comment (bparees) [inline pkg/generate/app/dockerimagelookup.go]: especially since they might go on to use the --allow-missing-images flag w/ their invalid argument.

**Comment by bparees** [comment]
@smarterclayton i think that's all i've got.

File: pkg/generate/app/dockerimagelookup.go (hunk 22 above)
Comment (smarterclayton) [inline pkg/generate/app/dockerimagelookup.go]: But we show you the typo.  Returning an error should actually be an error
that applies to the term - a term that can't be an image isn't really an
error.  "../foo/bar/template.json" isn't an error

**Comment by smarterclayton** [comment]
Squashed and rebased

File: pkg/generate/app/dockerimagelookup.go (hunk 22 above)
Comment (bparees) [inline pkg/generate/app/dockerimagelookup.go]: ok that makes sense, i wasn't considering the fact that this function is going to get called w/ args that were never intended to be matched to image specs.

**Comment by bparees** [comment]
lgtm

**Comment by smarterclayton** [comment]
[merge] and hope this fixes flakes

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test] - guess I just got really lucky.  oh well.
